DllImportGenerator/Microsoft.Interop.SourceGeneration/IGeneratorDiagnostics.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Interop
+{
+    public static class DiagnosticExtensions
+    {
+        public static Diagnostic CreateDiagnostic(
+            this ISymbol symbol,
+            DiagnosticDescriptor descriptor,
+            params object[] args)
+        {
+            IEnumerable<Location> locationsInSource = symbol.Locations.Where(l => l.IsInSource);
+            if (!locationsInSource.Any())
+                return Diagnostic.Create(descriptor, Location.None, args);
+
+            return Diagnostic.Create(
+                descriptor,
+                location: locationsInSource.First(),
+                additionalLocations: locationsInSource.Skip(1),
+                messageArgs: args);
+        }
+
+        public static Diagnostic CreateDiagnostic(
+            this AttributeData attributeData,
+            DiagnosticDescriptor descriptor,
+            params object[] args)
+        {
+            SyntaxReference? syntaxReference = attributeData.ApplicationSyntaxReference;
+            Location location = syntaxReference is not null
+                ? syntaxReference.GetSyntax().GetLocation()
+                : Location.None;
+
+            return Diagnostic.Create(
+                descriptor,
+                location: location.IsInSource ? location : Location.None,
+                messageArgs: args);
+        }
+    }
+
+    public interface IGeneratorDiagnostics
+    {
+        /// <summary>
+        /// Report diagnostic for marshalling of a parameter/return that is not supported
+        /// </summary>
+        /// <param name="method">Method with the parameter/return</param>
+        /// <param name="info">Type info for the parameter/return</param>
+        /// <param name="notSupportedDetails">[Optional] Specific reason for lack of support</param>
+        void ReportMarshallingNotSupported(
+            IMethodSymbol method,
+            TypePositionInfo info,
+            string? notSupportedDetails);
+
+        /// <summary>
+        /// Report diagnostic for configuration that is not supported by the DLL import source generator
+        /// </summary>
+        /// <param name="attributeData">Attribute specifying the unsupported configuration</param>
+        /// <param name="configurationName">Name of the configuration</param>
+        /// <param name="unsupportedValue">[Optiona] Unsupported configuration value</param>
+        void ReportConfigurationNotSupported(
```

Comment:
If this is a public API, I would prefer not to have default argument values and instead have method overrides.